### PR TITLE
Added entity view for Azure Repo Pull Requests

### PR DESCRIPTION
**Pull request by awanlin**
Signed-off-by: Andre Wanlin <awanlin@rapidrtc.com>

## Hey, I just made a Pull Request!

Added entity view for Azure Repo Pull Requests

azure-devops-pull-requests

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by awanlin** [comment]
@marleypowell here's the PR to add the entity view for Pull Requests, could you give it a look over, please? Most of the maintainers will be gone this week but I think that give me time to make adjustments to this after your review

File: catalog-info.yaml
Comment (freben) [inline catalog-info.yaml:21]: So is this unintentional or...? :)

**Review by freben — COMMENTED** [review]
Just a quick note. Also, I'm curious what it's like - a little screenshot in the PR description would be nice!

File: catalog-info.yaml
Comment (awanlin) [reply catalog-info.yaml]: Flipping between to many repos doing some testing, total accident, this has been reverted.

And I added a screenshot too 👍

File: plugins/azure-devops/src/components/BuildTable/BuildTable.tsx
Comment (marleypowell) [inline plugins/azure-devops/src/components/BuildTable/BuildTable.tsx:153]: This can probably be a common function in a seperate file as `getDurationFromDates` or something similar.

File: plugins/azure-devops/src/components/BuildTable/BuildTable.tsx
Comment (marleypowell) [inline plugins/azure-devops/src/components/BuildTable/BuildTable.tsx:118]:
```suggestion
export const getBuildDuration = (startTime?: Date, finishTime?: Date): string => {
```

File: plugins/azure-devops/src/components/BuildTable/BuildTable.test.ts
Comment (marleypowell) [inline plugins/azure-devops/src/components/BuildTable/BuildTable.test.ts:247]: You could add a test which checks the result for a valid start and end time duration.

File: plugins/azure-devops/src/components/BuildTable/BuildTable.test.ts
Comment (marleypowell) [inline plugins/azure-devops/src/components/BuildTable/BuildTable.test.ts:259]:
```suggestion
    it('should return empty result', () => {
```

File: plugins/azure-devops/src/components/BuildTable/BuildTable.test.ts
Comment (marleypowell) [inline plugins/azure-devops/src/components/BuildTable/BuildTable.test.ts:249]:
```suggestion
    it('should return empty result', () => {
```

File: plugins/azure-devops/src/components/AzureReposIcon/AzureReposIcon.tsx
Comment (marleypowell) [inline plugins/azure-devops/src/components/AzureReposIcon/AzureReposIcon.tsx:29]: Would this work?

```suggestion
const useStyles = makeStyles(theme => {
  icon: {
    filter: `invert(${theme.palette.type.toString() === 'dark' ? 100 : 0}%)`,
  },
});
```

File: plugins/azure-devops/src/components/AzureReposIcon/AzureReposIcon.tsx
Comment (marleypowell) [inline plugins/azure-devops/src/components/AzureReposIcon/AzureReposIcon.tsx:42]: MUI has an `SvgIcon` component which you might be able to use.
```ts
import StarIcon from './star.svg';

<SvgIcon component={StarIcon} viewBox="0 0 600 476.6" />
```

Also for the color of the icon would it work to use `fill="currentColor"` to automatically pick up the color?

Example of what I mentioned above:

```ts
import { SvgIcon, SvgIconProps } from '@material-ui/core';

import React from 'react';

export const AzurePullRequestsIcon = (props: SvgIconProps) => (
  <SvgIcon {...props} viewBox="0 0 512 512">
    <path
      d="M160 95.1c0-44.18-35.82-80-80-80c-44.18 0-80 35.82-80 80C0 128.8 19.77 156.9 48 169.2V342.8C19.77 355.1 0 383.2 0 415.1c0 44.18 35.82 80 80 80c44.18 0 80-35.82 80-80c0-32.79-19.77-60.89-48-73.25V169.2C140.2 156.9 160 128.8 160 95.1zM80 439.1c-13.23 0-24-10.77-24-24s10.77-24 24-24c13.23 0 24 10.77 24 24S93.23 439.1 80 439.1zM80 119.1c-13.23 0-24-10.77-24-24s10.77-24 24-24c13.23 0 24 10.77 24 24S93.23 119.1 80 119.1zM464 342.8v-182.8c0-52.94-43.06-96-96-96H320v-48c0-6.219-3.594-11.84-9.219-14.5c-5.594-2.594-12.25-1.781-17.03 2.219l-96 80C194.1 86.75 192 91.25 192 95.1S194.1 105.2 197.8 108.3l96 80C296.7 190.7 300.3 191.1 304 191.1c2.312 0 4.625-.5 6.781-1.5C316.4 187.8 320 182.2 320 175.1v-48h48c17.66 0 32 14.34 32 32v182.8c-28.23 12.36-48 40.46-48 73.25c0 44.18 35.82 80 80 80c44.18 0 80-35.82 80-80C512 383.2 492.2 355.1 464 342.8zM432 439.1c-13.23 0-24-10.77-24-24s10.77-24 24-24c13.23 0 24 10.77 24 24S445.2 439.1 432 439.1z"
      fill="currentColor"
    />
  </SvgIcon>
);
```

File: plugins/azure-devops/src/components/AzureReposIcon/AzureReposIcon.tsx
Comment (marleypowell) [inline plugins/azure-devops/src/components/AzureReposIcon/AzureReposIcon.tsx:31]: I think this is an `AzurePullRequestsIcon` as the `AzureReposIcon` looks slightly different.

File: plugins/azure-devops/src/components/AzurePipelinesIcon/AzurePipelinesIcon.tsx
Comment (marleypowell) [inline plugins/azure-devops/src/components/AzurePipelinesIcon/AzurePipelinesIcon.tsx:31]: Similar suggestions that I left on `AzureReposIcon` can also apply here.

File: plugins/azure-devops/src/api/AzureDevOpsClient.ts
Comment (marleypowell) [inline plugins/azure-devops/src/api/AzureDevOpsClient.ts:51]:
```suggestion
  public async getPullRequests(
```

**Review by marleypowell — APPROVED** [review]
Great PR 🙂. Just a few suggestions and nitpicks.

File: plugins/azure-devops/src/components/BuildTable/BuildTable.tsx
Comment (awanlin) [reply plugins/azure-devops/src/components/BuildTable/BuildTable.tsx]: 👍

File: plugins/azure-devops/src/components/BuildTable/BuildTable.test.ts
Comment (awanlin) [reply plugins/azure-devops/src/components/BuildTable/BuildTable.test.ts]: 👍

File: plugins/azure-devops/src/components/BuildTable/BuildTable.test.ts
Comment (awanlin) [reply plugins/azure-devops/src/components/BuildTable/BuildTable.test.ts]: 👍

File: plugins/azure-devops/src/api/AzureDevOpsClient.ts
Comment (awanlin) [reply plugins/azure-devops/src/api/AzureDevOpsClient.ts]: I've added this here and to the `getRepoBuilds` above to keep this consistent

File: plugins/azure-devops/src/components/AzureReposIcon/AzureReposIcon.tsx
Comment (awanlin) [reply plugins/azure-devops/src/components/AzureReposIcon/AzureReposIcon.tsx]: Probably but your suggestion to use `SvgIcon` proved to be the better solution

File: plugins/azure-devops/src/components/AzureReposIcon/AzureReposIcon.tsx
Comment (awanlin) [reply plugins/azure-devops/src/components/AzureReposIcon/AzureReposIcon.tsx]: Yes! This is so much nicer and I've used this. I was able to remove all that extra code and it's much leaner and cleaner :)

File: plugins/azure-devops/src/components/AzureReposIcon/AzureReposIcon.tsx
Comment (awanlin) [reply plugins/azure-devops/src/components/AzureReposIcon/AzureReposIcon.tsx]: I think this got fixed by the other changes

File: plugins/azure-devops/src/components/AzurePipelinesIcon/AzurePipelinesIcon.tsx
Comment (awanlin) [reply plugins/azure-devops/src/components/AzurePipelinesIcon/AzurePipelinesIcon.tsx]: Yup, fixed this the same way 👍

File: plugins/azure-devops/src/components/BuildTable/BuildTable.tsx
Comment (awanlin) [reply plugins/azure-devops/src/components/BuildTable/BuildTable.tsx]: Yes, I waffled on doing that. I've extracted it and the related tests

File: plugins/azure-devops/src/components/BuildTable/BuildTable.test.ts
Comment (awanlin) [reply plugins/azure-devops/src/components/BuildTable/BuildTable.test.ts]: So, I actually did but I had nothing but problems. Either the tests would fail or the changes that got the test working would cause the function to return `0s`. The issues I had was around using `fromISO` or `fromJSDate`. It was like what came from the Azure DevOps API I could not replicate in the test. 

How does this sound: let's get this approved and then I'll submit a PR with the tests and we can work together to figure out what I'm doing wrong. I do like having tests and I'm happy to put in the work on this.

**Review by marleypowell — APPROVED** [review]
LGTM 🙂

File: plugins/azure-devops/src/components/PullRequestTable/PullRequestTable.tsx
Comment (marleypowell) [inline plugins/azure-devops/src/components/PullRequestTable/PullRequestTable.tsx:83]:
```suggestion
      DateTime.fromJSDate(row.creationDate ?? new Date()).toRelative(),
```

Comment (awanlin) [reply plugins/azure-devops/src/components/PullRequestTable/PullRequestTable.tsx]: This has been fixed 👍

**Comment by awanlin** [comment]
@freben this is now in a place where you or any of the other maintainers can take a look when you have time, please :)

**Comment by marleypowell** [comment]
@awanlin FYI #7672 may conflict with this PR as it's changing the location of some of the types that you've updated. Are you okay for my PR to merge first?

File: packages/app/src/components/catalog/EntityPage.tsx
Comment (Rugvip) [inline packages/app/src/components/catalog/EntityPage.tsx:271]: Slight nit on the naming here, perhaps `EntityAzurePullRequestsContent` for clarity?

**Review by Rugvip — COMMENTED** [review]
Thank you! 👍 

Wiring looks good, didn't dive too deep into the implementation, trusting you with that :grin:

Just a slight naming nit to consider

File: plugins/azure-devops/src/components/PullRequestTable/PullRequestTable.tsx
Comment (marleypowell) [inline plugins/azure-devops/src/components/PullRequestTable/PullRequestTable.tsx:47]:
```suggestion
        <Link to={row.link ?? ''}>{row.title}</Link>
```

File: plugins/azure-devops/src/components/PullRequestTable/PullRequestTable.tsx
Comment (marleypowell) [inline plugins/azure-devops/src/components/PullRequestTable/PullRequestTable.tsx:48]: Should this be inside the `row.isDraft` condition?

**Review by marleypowell — APPROVED** [review]
A few small nit picks.

File: packages/app/src/components/catalog/EntityPage.tsx
Comment (awanlin) [reply packages/app/src/components/catalog/EntityPage.tsx]: Thanks for taking a look. Yes, I was holding on to this naming but Marley mention another place he found it confusing and I reconsidered my stance and have adjusted this to be more clear. 👍

File: plugins/azure-devops/src/components/PullRequestTable/PullRequestTable.tsx
Comment (awanlin) [reply plugins/azure-devops/src/components/PullRequestTable/PullRequestTable.tsx]: I've made this adjustment 👍

File: plugins/azure-devops/src/components/PullRequestTable/PullRequestTable.tsx
Comment (awanlin) [reply plugins/azure-devops/src/components/PullRequestTable/PullRequestTable.tsx]: Yes, good catch and fixed!

**Comment by awanlin** [comment]
@marleypowell @Rugvip I've rebased to get the changes to use the new `azure-devops-common` and also the suggestions that were made recently. Can you take another look? Would be cool to get this into the next release if possible 😎

File: plugins/azure-devops/src/components/PullRequestStatusButtonGroup/PullRequestStatusButtonGroup.tsx
Comment (marleypowell) [inline plugins/azure-devops/src/components/PullRequestStatusButtonGroup/PullRequestStatusButtonGroup.tsx:27]:
```suggestion
  setPullRequestsStatusState: (pullRequestStatus: PullRequestStatus) => void;
```

File: plugins/azure-devops/src/components/PullRequestTable/PullRequestTable.tsx
Comment (marleypowell) [inline plugins/azure-devops/src/components/PullRequestTable/PullRequestTable.tsx:84]: I think we need to be converting dates to ISO strings before returning them in API responses.

When we deserialize an API response they are typed as `Date` but they're actully still `string`.

File: plugins/azure-devops/src/components/PullRequestStatusButtonGroup/PullRequestStatusButtonGroup.tsx
Comment (awanlin) [reply plugins/azure-devops/src/components/PullRequestStatusButtonGroup/PullRequestStatusButtonGroup.tsx]: Fixed with proper type, thanks

File: plugins/azure-devops/src/components/PullRequestTable/PullRequestTable.tsx
Comment (awanlin) [reply plugins/azure-devops/src/components/PullRequestTable/PullRequestTable.tsx:84]: As this will require backend changes I'll hold off on this PR until I can get those changes in

**Comment by awanlin** [comment]
@freben @Rugvip please hold off on this PR, @marleypowell has pointed out a larger issue I need to fix. Sorry for all the noise :(

File: plugins/azure-devops/src/plugin.ts
Comment (marleypowell) [inline plugins/azure-devops/src/plugin.ts:67]: Does this need to have a different route ref? Also the naming could maybe be slightly more specific.

```
export const azurePipelinesEntityContentRouteRef = createRouteRef({
  title: 'Azure Pipelines Entity Content',
});

export const azurePullRequestsEntityContentRouteRef = createRouteRef({
  title: 'Azure Pull Requests Entity Content',
});
```

Comment (awanlin) [reply plugins/azure-devops/src/plugin.ts:67]: Yes, I'll make these changes, this part is still a little magical to me. I also like the naming, it's more clear, thanks

Comment (awanlin) [reply plugins/azure-devops/src/plugin.ts]: These changes have been made, did some minor clean up in the `plugin.ts` and `routes.ts` files as well

**Review by marleypowell — APPROVED** [review]

File: plugins/azure-devops/src/components/PullRequestTable/PullRequestTable.tsx
Comment (awanlin) [reply plugins/azure-devops/src/components/PullRequestTable/PullRequestTable.tsx]: These changes have been made and dates are handled much more nicely now

File: plugins/azure-devops/src/components/BuildTable/BuildTable.test.ts
Comment (awanlin) [reply plugins/azure-devops/src/components/BuildTable/BuildTable.test.ts]: With the changes to improve date handling I've added the test in that I was having issues with before so this is now tested properly 👍

**Comment by awanlin** [comment]
Alright, ready for another review @marleypowell, @freben, @Rugvip

**Review by marleypowell — APPROVED** [review]

File: plugins/azure-devops/src/api/AzureDevOpsClient.ts
Comment (freben) [inline plugins/azure-devops/src/api/AzureDevOpsClient.ts:57]: Whenever forming a URL, you should do `encodeURIComponent` on the string so that a malicious user can't craft URLs that break the patterns and worst case cause trouble.

I see that the `top` and `status` fieds are optional, but still are added unconditionally to the URL which probably means that the string "undefined" will be in the URL. Might I suggest doing something like this?

```ts
const q = new URLSearchParams();
if (options?.top) {
  q.append('top', options.top);
}
if (options?.status) {
  q.append('status', options.status;
}
const url = `pull-requests/${encodeURIComponent(projectName)}/${encodeURIComponent(repoName)}?${q}`
```

(both of these apply to the other method above too of course).

File: plugins/azure-devops/src/components/EntityPageAzurePullRequests/EntityPageAzurePullRequests.tsx
Comment (freben) [inline plugins/azure-devops/src/components/EntityPageAzurePullRequests/EntityPageAzurePullRequests.tsx:25]: seems like a useless wrapper component at this point - maybe there will be more usages later on?

File: plugins/azure-devops/src/components/PullRequestStatusButtonGroup/PullRequestStatusButtonGroup.tsx
Comment (freben) [inline plugins/azure-devops/src/components/PullRequestStatusButtonGroup/PullRequestStatusButtonGroup.tsx:24]: Just a matter of taste, but maybe it can be argued that these are overly verbose given that you already express the full context in the component name. `<PullRequestStatusButtonGroup status={} setStatus={}>` might be enough

File: plugins/azure-devops/src/hooks/usePullRequests.ts
Comment (freben) [inline plugins/azure-devops/src/hooks/usePullRequests.ts:51]: should probably make this `[api, project, repo, top, status]` and place the `options` only inside the `useAsync`

**Review by freben — COMMENTED** [review]
Nice! Just some small things

File: plugins/azure-devops/src/routes.ts
Comment (benjdlambert) [inline plugins/azure-devops/src/routes.ts:20]: Just a small nit here, when this lands https://github.com/backstage/backstage/pull/7702, `title` is not longer set by the `routeRef` and `id` becomes mandatory. Is this something we could incorporate into this change too see as your playing around with them anyway?

File: plugins/azure-devops/src/hooks/usePullRequests.ts
Comment (marleypowell) [inline plugins/azure-devops/src/hooks/usePullRequests.ts:43]:
```suggestion
    top,
    status,
```

Comment (awanlin) [reply plugins/azure-devops/src/hooks/usePullRequests.ts]: 👍

File: plugins/azure-devops/src/routes.ts
Comment (awanlin) [reply plugins/azure-devops/src/routes.ts]: For sure, I've made this change, thanks

Comment (awanlin) [reply plugins/azure-devops/src/routes.ts]: @marleypowell you might want to do this as well

File: plugins/azure-devops/src/hooks/usePullRequests.ts
Comment (awanlin) [reply plugins/azure-devops/src/hooks/usePullRequests.ts]: Sounds good, this change has been made

File: plugins/azure-devops/src/routes.ts
Comment (marleypowell) [reply plugins/azure-devops/src/routes.ts]: > @marleypowell you might want to do this as well

Thanks I will make sure to include it in my PR.

File: plugins/azure-devops/src/components/PullRequestStatusButtonGroup/PullRequestStatusButtonGroup.tsx
Comment (awanlin) [reply plugins/azure-devops/src/components/PullRequestStatusButtonGroup/PullRequestStatusButtonGroup.tsx]: This is a little nicer so I've gone this route, thanks

File: plugins/azure-devops/src/components/EntityPageAzurePullRequests/EntityPageAzurePullRequests.tsx
Comment (awanlin) [reply plugins/azure-devops/src/components/EntityPageAzurePullRequests/EntityPageAzurePullRequests.tsx:25]: Yes, I have some ideas here so let's leave it as is for now, thanks

File: plugins/azure-devops/src/api/AzureDevOpsClient.ts
Comment (awanlin) [reply plugins/azure-devops/src/api/AzureDevOpsClient.ts]: Yes, this is a big gap in my knowledge so thanks for pointing this out. Both have been updated with this and I'll know to do this in the future as well 👍

Comment (freben) [reply plugins/azure-devops/src/api/AzureDevOpsClient.ts]: Nice! Thanks

**Comment by awanlin** [comment]
Thanks for the reviews @freben, @benjdlambert, and @marleypowell! I've made the changes could you take another look, please? Thanks in advance! :)

**Review by marleypowell — APPROVED** [review]

**Review by freben — APPROVED** [review]
Looks good to me!

**Pull request merged**
